libs/core/build.gradle

```diff
@@ -79,6 +79,9 @@ dependencies {
 
   api "com.fasterxml.jackson.core:jackson-core:${versions.jackson}"
 
+  // lucene
+  api "org.apache.lucene:lucene-core:${versions.lucene}"
```

Comment:
@nknize I have doubts we need to depend on Lucene in any core / common modules, the `BytesRefUtils` is very specialized utility class, why do we need to promote it to core?

Reply:
Per #5910. Consider a plugin that has it's own custom Analyzer using `AnalysisPlugin`. Downstream plugins implement a `TokenFilterFactory` which needs Lucene's `TokenStream`.  We need to take a dependency on lucene for this in whatever library houses those contracts (currently decided to be the `opensearch-core` library).

Reply:
Separately from that, I could remove `BytesRefUtils` altogether as that logic isn't actually used anywhere. I chose to keep it (for now) as a convenience function for external plugins.

Reply:
> We need to take a dependency on lucene for this in whatever library houses those contracts (currently decided to be the opensearch-core library).

This is totally fine, and I believe by importing `AnalysisPlugin` the Lucene will be brought in, but the core does not need it and should not bring it I think. Take libsfor example - they hardly ever need to know about Lucene but very likely will depend on core / common.

> Separately from that, I could remove BytesRefUtils altogether as that logic isn't actually used anywhere.

👍 ty

Reply:
>  ...but the core does not need it and should not bring it I think

I think we're not on the same page. #5910 plans to move general contracts like `AnalysisPlugin` (and all of the others) out of `:server` and into `opensearch-core` library so plugins only need take a dependency on the API not the entire `:server` module. To do that requires classes like `*Plugin`, be refactored out of `:server` and into `libs:opensearch-core`. If you refactor `AnalysisPlugin` to `:opensearch-core` you need `:opensearch-core` to take a dependency on lucene. So, the core library will need this lucene dependency, otherwise these contracts have to stay in `:server`

Reply:
So maybe `AnalysisPlugin` goes into an `:opensearch-analysis` library. Take another class, `Version.java`. This should be in the `opensearch-core` library, and it ties to Lucene's `Version.java`, so `:opensearch-core` would need a dependency on `lucene-core`.  `OpenSearchException` is another core class... While I'm untangling a lot of the server classes, it still will need a dependency on `lucene-core`. 

The point I'm making here is that `:opensearch-common` doesn't take a dependency on lucene because it's general java collections. But OpenSearch is built around the Lucene API. Unless we want to abstract the Lucene API into a general search library, why should we be afraid for `opensearch-core` to take a dependency on Lucene? `:libs:opensearch-core` should be more about isolating the concrete server implementation from the plugins, not the entire Lucene API.

Reply:
Thanks @nknize , I am not against the vision but the execution is lacking behind I think. We need to rebalance the `common` / `core` modules to preserve the dependencies constraints. Just one of the examples I shared previously, `opensearch-nio`, this module depends on :+1: 

```
dependencies {
  api project(':libs:opensearch-core')
  api project(':libs:opensearch-common')
  ...
}
```

And has nothing to do nor with Lucene, nor with plugins etc. But by this simple change in core,we would introduce a cascading effect of unnecessary transitive dependencies, this should not happen. I am sure there are many more examples of similar impact.

The other issue is that plugins vs extensions. Since we intend to bundle everything in core, how does it look developers who do plugins vs extensions? Do we need to think about splitting these by need for example? Should as an extension developer I care about plugins / Lucene / etc, similarly as plugin developer, should I care about extensions / protobuf (coming) / etc?

Reply:
> Do you forsee a scenario where opensearch-nio would be used in an application without Lucene? 

Absolutely, to be fair I don't know how reusable this particular part is, but Lucene is out of scope for it. I mean this is just the first example I run into - you could look, there are more (in `libs` modules). 

> Are you suggesting have another dependency free library (e.g., no lucene) with OpenSearch contracts that don't have any third-party (e.g., lucene) dependencies?

At this moment - no, I am suggesting to rebalance the `common` / `core` modules. Like why `libs/opensearch-nio` needs both modules, may be we could move some types from `core` to `common` and drop the dependency?

> What do you mean by this?

I should have been more clear, I meant  bundle everything related to plugins and extension in core. So no matter what you want to develop - you will get both (alternatively, thinking out loud, having `opensearch-plugins-spi` and ``opensearch-extensions-spi` split).

> So my big concern here is that we need a decision. Are we going to prevent opensearch-core from having any third party dependencies? 

No, we have `opensearch-common` for that, we should prevent the dependency on  `opensearch-core` where it is not necessary.

Reply:
> No, we have opensearch-common for that, we should prevent the dependency on opensearch-core where it is not necessary.

I agree with this. Bringing Lucene in as a dependency on `opensearch-core` in this PR does not prevent us from working toward that end. So let's separate that goal from this PR?

Reply:
> why libs/opensearch-nio needs both modules

It only needs both because of it's dependency on [NetUtils](https://github.com/opensearch-project/OpenSearch/blob/main/libs/core/src/main/java/org/opensearch/core/internal/net/NetUtils.java) we can refactor that class out of core in a separate PR if we'd like so we're not conflating that with this change here. We could put it in common but, as I mentioned before, common is used by `build-tools` and we'll be opening pandora's box for build-tools pulling in a lot of unnecessary dependencies. Maybe we take your [earlier suggestion](https://github.com/opensearch-project/OpenSearch/pull/5976#issuecomment-1403967166) and create an `opensearch-buildtools` library only for plugins taking a dependency on `opensearch-build-tools`? Do that in a separate PR as well?

Reply:
So here's what I suggest... 

1. I don't see a problem this PR pulling a dependency on Lucene, so we're safe to merge this.
2. Address the `common` `core` library entaglement separately (e.g., refactor NetUtils to common)
3. Address the `build-tools` dependency separately (e.g., create a new `opensearch-build` library for `build-tools` and devops specific library implementation)

Reply:
> I don't see a problem this PR pulling a dependency on Lucene, so we're safe to merge this.

I have run into unwanted dependency mess too many times (personal experience), I assume that you want to bring this to `2.x` and I don't want to surprise people with new transitive dependencies coming in. I would not take such decisions lightly (again, just personal experience)

Reply:
> I could help you do the pre-work

Pre-work for what?

> moreover I think you agreed that the case is not made for it - ByteRefUtils could live in other place (for now).

For this PR, but the next PR is about to refactor `Version.java` to core library which is a hard dependency on `lucene.Version.java`. Everything is in server! Every plugin already takes a dependency on server which is 29 transitive dependencies, so I'm not sure why you're concerned about transitive dependencies that already exist today? Furthermore, nio can go away! It was only there for imitations with legacy java nio (I think elasticsearch already removed this). So I'm not concerned about that library. 


> I assume that you want to bring this to 2.x and I don't want to surprise people with new transitive dependencies coming in.


Yes. Because it moves 2.x toward reducing the dependency on :server which is much more dangerous then pulling in Lucene dependencies that already exist.

Reply:
> Pre-work for what?

Remove dependency from `opensearch-core` where `opensearch-common` would be sufficient (possibly moving few classes between modules).

> What OpenSearch library will cause this that's not already used along with opensearch proper in a downstream consumer? Put another way. Is this just FUD due to past history?

Up to now, OpenSearch (and ElasticSearch in general) were pretty stable with respect to modules and dependencies, we are changing that. It is not FUD, what I am afraid of, from less practical `libs` to more practical, are clients: they are used everywhere, RHLC is a great example of spitting dependencies all over the place.

Reply:
> are clients: they are used everywhere, RHLC is a great example of spitting dependencies all over the place.

RHLC already depends on :server. :rest is unaffected because it doesn't depend on :core or :server, :benchmark already depends on :server, :client-benchmark-noop-api-plugin doesn't depend on either. opensearch-java already depends on RHLC which (as above) depends on :server so it will benefit from this refactor (not exposing all of server). So this doesn't affect clients negatively at all. If anything this move away from server will only reduce client transitive dependencies by way of reducing the dependency on server. 

But we can't get there without refactoring the general server components into the core library.

Reply:
As another example, look at `Version.java`.  In order for downstream plugins to provide bwc they have to use the `Version.onOrBefore(Version.2_6_0)` logic to ensure non breaking code paths. Is this the way we want it in OpenSearch 3x, 4x, 10x? No.. we want semver handled through some other actual semver mechanism. But this is the reality we're in today... So, what CAN we do to remove the downstream dependency on :server so we can protect classes like `InternalEngine` without breaking 2.x? Refactor Version.java to `opensearch-core` library in `org.opensearch.core.Version` and replace the downstream dependency on :server with the lighter `opensearch-core` `opensearch-common` (and whatever third party transitive dependencies).

We're blocking that today out of concern for transitive dependencies THAT ARE ALREADY EXISTING

Reply:
> But we can't get there without refactoring the general server components into the core library.

I am :100: with you here, that is why cautions about dependencies  - we should not repeat `server` story (even at smaller scale).

> We're blocking that today out of concern for transitive dependencies THAT ARE ALREADY EXISTING

There is no blocking (at least from my side), just suggesting to take different route. For some modules transitive dependencies are already existing (will be cleaned at some point), for some not (I think this is good thing). 

@andrross we probably need your fresh look in here :D

Reply:
> What's the different route?

First thing, remove dependency from `opensearch-core` where `opensearch-common` would be sufficient (possibly moving few classes between modules). With that - we could be 1000% sure that no new transitive dependencies would surprise us along the away.

Reply:
In the meantime I updated this PR to remove `opensearch-core` dependency from `ssl-config` and `nio` libraries.

Reply:
@nknize `libs/opensearch-cli` ?

Reply:
+1 done

Reply:
I was hoping all of this would be done incrementally after this PR was merged.

There is an argument for moving the xcontent classes out of core into the common library now and letting core take the dependency on Lucene and other third-party libraries. Which would mean all of the plugins would revert the PRs they just merged. This is why we need to decouple downstream repos to enable us to move on core changes.